File: backend/app/services/slice_formats.py
"""Single source of truth for which model formats the slicer sidecar accepts.

STEP/STP are deliberately absent: the OrcaSlicer/BambuStudio CLI behind the
sidecar cannot import them reliably, so offering a server-side slice for a
STEP file only produces a job that fails minutes later. Uploading, storing and
previewing STEP files stays supported — see issue #92.
"""

from __future__ import annotations

# Model geometry inputs the sidecar can slice. `.gcode.3mf` also matches here
# on purpose: re-slicing an already-sliced 3MF is a supported flow.
SERVER_SLICEABLE_EXTENSIONS: tuple[str, ...] = (".stl", ".3mf")

# Formats we accept for upload/library management but cannot slice server-side.
UNSLICEABLE_CAD_EXTENSIONS: tuple[str, ...] = (".step", ".stp")

SLICEABLE_FORMATS_LABEL = "STL or 3MF"


def is_server_sliceable_filename(filename: str | None) -> bool:
    """Return True when ``filename`` can be fed to the slicer sidecar."""
    return (filename or "").lower().endswith(SERVER_SLICEABLE_EXTENSIONS)


def is_unsliceable_cad_filename(filename: str | None) -> bool:
    """Return True for CAD formats we store but cannot slice (STEP/STP)."""
    return (filename or "").lower().endswith(UNSLICEABLE_CAD_EXTENSIONS)


def unsliceable_detail(filename: str | None, *, subject: str = "Source file") -> str:
    """Human-readable 400 detail explaining why a slice was refused."""
    if is_unsliceable_cad_filename(filename):
        return (
            f"{subject} is a STEP/STP file, which the slicer sidecar cannot "
            "process. Convert it to STL or 3MF and slice that instead."
        )
    return f"{subject} must be {SLICEABLE_FORMATS_LABEL} to slice"

File: frontend/src/utils/sliceFormats.ts
/**
 * Which model formats PrintOps can slice server-side.
 *
 * STEP/STP are deliberately excluded: the OrcaSlicer/BambuStudio CLI behind
 * the slicer sidecar cannot import them reliably, so offering the action only
 * produces a job that fails later. Upload, library management and 3D preview
 * of STEP files stay untouched. Mirrors
 * `backend/app/services/slice_formats.py` — see issue #92.
 */

const SLICEABLE_EXTENSIONS = ['.stl', '.3mf'] as const;
const UNSLICEABLE_CAD_EXTENSIONS = ['.step', '.stp'] as const;

/** True when the sidecar can slice this file (model geometry, not G-code). */
export function isServerSliceableFilename(filename: string | null | undefined): boolean {
  const lower = (filename || '').toLowerCase();
  if (lower.endsWith('.gcode') || lower.endsWith('.gcode.3mf')) return false;
  return SLICEABLE_EXTENSIONS.some((ext) => lower.endsWith(ext));
}

/** True for CAD formats PrintOps stores but cannot slice server-side. */
export function isUnsliceableCadFilename(filename: string | null | undefined): boolean {
  const lower = (filename || '').toLowerCase();
  return UNSLICEABLE_CAD_EXTENSIONS.some((ext) => lower.endsWith(ext));
}

/** Same check keyed off a `file_type` value (`'stl'`, `'step'`, …). */
export function isServerSliceableFileType(fileType: string | null | undefined): boolean {
  const t = (fileType || '').toLowerCase();
  return t === 'stl' || t === '3mf';
}

/** True when a `file_type` value names a CAD format we cannot slice. */
export function isUnsliceableCadFileType(fileType: string | null | undefined): boolean {
  const t = (fileType || '').toLowerCase();
  return t === 'step' || t === 'stp';
}
